README.md
# Sliding Card

![](https://media.giphy.com/media/d8uClfpq4R6jKQvV8T/giphy.gif)

## Introduction

Sliding card is a highly customizable flutter package that will help you create beautiful Cards with a sliding animation effect.

## Before we start

Before you start here are some relevant link that you might want to have a look

* [The online documentation](https://www.emilecode.com/sliding-card/ "Sliding Card")
* [An UI made using Sliding card](https://github.com/PavieOlivier/doctor-appointment-flutter "Sliding Card UI")
* [A Demo APK](https://drive.google.com/open?id=1ymME7ZY9VGNXdhPprEz0hz4l915lLkp1 "Sliding Card Demo APK")
* [My instagram account](https://www.instagram.com/emilecode/ "emilecode")

## Quick start

To use the application just import the package 
```
import 'package:sliding_card/sliding_card.dart';
```
Then inside your code, add the SlidingCard widget.
```
SlidingCard(
        slimeCardElevation: 0.5,
        cardsGap: SizeConfig.safeBlockVertical,
        //The controller is necessary to animate the opening and closing of the card
        controller: widget.slidingCardController,
        slidingCardWidth: SizeConfig.horizontalBloc * 90,
        visibleCardHeight: SizeConfig.safeBlockVertical * 27,
        //the card behind the front card can't have a higher
        // hight than the front one  
        hiddenCardHeight: SizeConfig.safeBlockVertical * 15,
        //Configure your front card here
        frontCardWidget: TheFrontCard(),
        //configure your rear card here 
        backCardWidget:TheBackCard(),
      ),
```

## Conclusion 
Thank you for using my package, If you did find any bug please feel free to report it or make a pull request.
If you want to contact me directly, please DM me directly on [My instagram account](https://www.instagram.com/emilecode/ "emilecode"), 
or you can mail me directly.
For my other packages you can visit my website at [emilecode.com](emilecode.com)



CHANGELOG.md
## [0.0.1] - Closed Beta.

* TODO: Released in close beta

## [0.1.0] - Initial public release day.

* Release on pub.dev 

## [0.1.1] - Initial public release day fix.

* uploaded readme.md

## [0.1.2] - Initial public release day fix.

* Re-formated package


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:sliding_card/sliding_card.dart';
import 'sizeConfig.dart';
import 'AppointmentCard.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Sliding Card'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
SlidingCardController controller ;
  @override
  void initState() {
    
    super.initState();
    controller = SlidingCardController();
  }
  @override
  Widget build(BuildContext context) {
    SizeConfig().init(context);
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            AppointmentCard(
              onCardTapped: () {
                print('Card tapped controller can be used here ');
                if(controller.isCardSeparated == true)
                {
                  controller.collapseCard();
                }
                else
                {
                  controller.expandCard();
                }
              },
              slidingCardController: controller,
            )
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/sliding_card.dart
library sliding_card;

import 'package:flutter/material.dart';

///This Class is used to control the opening and
///closing of the card
class SlidingCardController {
  ///This is to slide the card down
  Function expandCard;

  ///This is to slide the card up
  Function collapseCard;

  ///Whether the card is collapsed or not
  bool isCardSeparated;
}

///THIS IS A WORK IN PROGRESS
class SlidingCard extends StatefulWidget {
  final SlidingCardController controller;

  ///Widget to be displayed on the front;
  ///note that It's parent widget is a container
  final Widget frontCardWidget;

  ///Widget to be displayed on the back
  ///It's hight shhall be <= height of the front card;
  ///Note that it's parent widget is a container
  final Widget backCardWidget;

  ///Height of the front card
  final double visibleCardHeight;

  ///height of the back card,It
  ///cannot be greater than the height of the front card
  final double hiddenCardHeight;

  ///width of the overal card
  final double slidingCardWidth;

  ///Border radious of the card
  final double slimeCardBorderRadius;

  ///The elevation of the SlimeCard
  final double slimeCardElevation;

  ///The time it will take for the card to slide down and up
  final Duration slidingAnimmationDuration;

  ///The curve of the slidingAnimation when expanding
  ///it is preferable to leave it to it's default value
  final Curve slidingAnimmationForwardCurve;

  ///The curve of the slidingAnimation when reserving
  ///it is preferable to leave it to it's default value
  final Curve slidingAnimationReverseCurve;

  ///This gives a good visual effect
  ///Leave this to true for a more realistic
  final bool animateOpacity;

  /// this will enable coloring of the cards
  /// use this to make sure your widgets are aligned properly
  final bool showColors;

  ///The space between the 2 cards
  final double cardsGap;
  SlidingCard(
      {this.cardsGap = 12,
      this.slimeCardBorderRadius = 25,
      this.slidingCardWidth = 150,
      this.visibleCardHeight = 100,
      this.hiddenCardHeight = 100,
      this.animateOpacity = true,
      this.slimeCardElevation = 1,
      this.showColors = true,
      this.slidingAnimmationForwardCurve = Curves.elasticOut,
      this.slidingAnimationReverseCurve = Curves.elasticOut,
      this.slidingAnimmationDuration = const Duration(milliseconds: 800),
      @required this.controller,
      @required this.backCardWidget,
      @required this.frontCardWidget})
      : assert(hiddenCardHeight <= (visibleCardHeight * 90) / 100,
            'The height of the hidden card shall be less than or equal to 90% of the frontCard Widget ');

  @override
  _SlidingCardState createState() => _SlidingCardState();
}

class _SlidingCardState extends State<SlidingCard>
    with TickerProviderStateMixin {
  AnimationController animationController;
  Tween<double> separatedValue;
  Animation animationValue;
  bool isSeparated = false;
  Curve animationCurve;

  @override
  void initState() {
    super.initState();
    if (widget.controller != null) {
      widget.controller.expandCard = expandCard;
      widget.controller.collapseCard = collapseCard;
      widget.controller.isCardSeparated = false;
    } else {
      print(
          'Sliding Card: The controller is null, the sliding card action will not be triggered');
    }
    animationCurve = widget.slidingAnimmationForwardCurve;
    animationController =
        AnimationController(vsync: this, duration: Duration(milliseconds: 250));
    separatedValue =
        Tween(begin: 0.0, end: widget.visibleCardHeight + widget.cardsGap);
    animationValue = separatedValue.animate(animationController);
    animationController.addListener(() {
      setState(() {});
    });
  }

  ///use this to expand the card
  void expandCard() {
    //well we have to repeat this peace of code because it looks like in
    //release mode , the app is too fast and the tween does not get registered fast
    separatedValue =
        Tween(begin: 0.0, end: widget.visibleCardHeight + widget.cardsGap);
    animationCurve = widget.slidingAnimmationForwardCurve;
    animationValue = separatedValue.animate(animationController);
    animationController.forward();
    isSeparated = true;
    widget.controller.isCardSeparated = true;
  }

  ///use this to colapse the card
  void collapseCard() {
    animationCurve = widget.slidingAnimationReverseCurve;
    animationController.reverse();
    isSeparated = false;
    widget.controller.isCardSeparated = false;
  }

  @override
  void dispose() {
    super.dispose();
    animationController.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedContainer(
      //color: Colors.limeAccent,
      curve: animationCurve,
      duration: widget.slidingAnimmationDuration,
      height: isSeparated
          ? (widget.visibleCardHeight + widget.hiddenCardHeight) +
              (MediaQuery.of(context).size.height / 100) * 1.5
          : widget.visibleCardHeight,
      // color: Colors.red,
      child: Stack(
        children: <Widget>[
          //this is the back card
          Transform.translate(
              offset: Offset(0, animationValue.value),
              child: AnimatedOpacity(
                duration: isSeparated
                    ? Duration(milliseconds: 400)
                    : Duration(milliseconds: 800),
                opacity: widget.animateOpacity ? isSeparated ? 1 : 0 : 1,
                child: Card(
                  elevation: widget.slimeCardElevation,
                  shape: RoundedRectangleBorder(
                      borderRadius:
                          BorderRadius.circular(widget.slimeCardBorderRadius)),
                  child: Container(
                    width: widget.slidingCardWidth,
                    height: widget.hiddenCardHeight,
                    decoration: BoxDecoration(
                        color: widget.showColors ? Colors.yellow : null,
                        borderRadius: BorderRadius.circular(
                            widget.slimeCardBorderRadius)),
                    child: widget.backCardWidget,
                  ),
                ),
              )),

          //this is the front card
          Card(
            elevation: widget.slimeCardElevation,
            shape: RoundedRectangleBorder(
                borderRadius:
                    BorderRadius.circular(widget.slimeCardBorderRadius)),
            child: Container(
              width: widget.slidingCardWidth,
              height: widget.visibleCardHeight,
              decoration: BoxDecoration(
                  color: widget.showColors ? Colors.green : null,
                  borderRadius:
                      BorderRadius.circular(widget.slimeCardBorderRadius)),
              child: widget.frontCardWidget,
            ),
          ),
        ],
      ),
    );
  }
}


example/lib/AppointmentBackCard.dart
import 'sizeConfig.dart';
import 'package:flutter/material.dart';

class AppointmentBackCard extends StatelessWidget {
  final String patientComment;
  final Function onPhoneTapped;
  const AppointmentBackCard({
    Key key,
    @required this.patientComment,
    @required this.onPhoneTapped,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Column(
      children: <Widget>[
        Container(
          width: SizeConfig.horizontalBloc * 90,
          height: SizeConfig.safeBlockVertical * 15,
          decoration: BoxDecoration(
              color: Colors.white, borderRadius: BorderRadius.circular(25)),
          child: Padding(
            padding: const EdgeInsets.only(left: 20, right: 20, top: 8),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: <Widget>[
                Expanded(
                    child: Padding(
                  padding: const EdgeInsets.only(top: 0.0),
                  child: Text(
                    'Comment :',
                    style: TextStyle(
                        fontSize: SizeConfig.safeBlockHorizontal * 5,
                        fontWeight: FontWeight.bold,
                        color: Colors.blue[300],),
                  ),
                )),
                Expanded(
                  flex: 4,
                  child: Padding(
                    padding: const EdgeInsets.only(bottom: 8.0),
                    child: Container(
                      width: SizeConfig.safeBlockHorizontal * 80,
                      child: Row(
                        children: <Widget>[
                          Expanded(
                            flex: 3,
                            child: Container(
                              child: Text(
                                patientComment,
                                style: TextStyle(
                                    fontSize:
                                        SizeConfig.safeBlockHorizontal * 4,
                                    color: Colors.black),
                                overflow: TextOverflow.ellipsis,
                                maxLines: 4,
                              ),
                            ),
                          ),
                          Expanded(
                              child: Padding(
                            padding: const EdgeInsets.all(8.0),
                            child: InkWell(
                              onTap: onPhoneTapped,
                              child: Container(
                                decoration: BoxDecoration(
                                    color: Colors.blue[300],
                                    shape: BoxShape.circle),
                                child: Center(
                                    child: Icon(
                                  Icons.phone,
                                  size: SizeConfig.safeBlockHorizontal * 9,
                                  color: Colors.white,
                                )),
                              ),
                            ),
                          ))
                        ],
                      ),
                    ),
                  ),
                ),
              ],
            ),
          ),
        )
      ],
    );
  }
}


example/lib/AppointmentFrontCard.dart
import 'package:flutter/material.dart';
import 'sizeConfig.dart';

class AppointmentFrontCard extends StatefulWidget {
  final String patienName;
  final String patienSurname;
  final Function onInfoTapped;
  final Function onRedCloseButtonTapped;
  final Function onAccep;
  final Function onDecline;
  final String appointmentDate;
  final String appointmentTime;
  final String imgLink;
  const AppointmentFrontCard({
    Key key,
    @required this.imgLink,
    @required this.onAccep,
    @required this.onDecline,
    @required this.patienName,
    @required this.onInfoTapped,
    @required this.patienSurname,
    @required this.appointmentDate,
    @required this.appointmentTime,
    @required this.onRedCloseButtonTapped
  }) : super(key: key);
  @override
  _AppointmentFrontCardState createState() => _AppointmentFrontCardState();
}

class _AppointmentFrontCardState extends State<AppointmentFrontCard> {
  bool isinfoPressed = false;
  @override
  Widget build(BuildContext context) {
    return Column(
      children: <Widget>[
        Expanded(
          child: Container(
              decoration: BoxDecoration(
                  color: Color(0xff2C8BFF),
                  borderRadius: BorderRadius.only(
                      topLeft: Radius.circular(25),
                      topRight: Radius.circular(25))),
              child: Padding(
                padding: const EdgeInsets.only(top: 0, left: 20),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: <Widget>[
                    Expanded(
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.spaceBetween,
                        children: <Widget>[
                          Padding(
                            padding: const EdgeInsets.only(top: 8.0),
                            child: Text(
                              'Appointment Request',
                              style: TextStyle(
                                  fontSize: SizeConfig.safeBlockHorizontal * 4.5,
                                  color: Colors.white),
                            ),
                          ),
                          IconButton(
                            onPressed: () {},
                            icon: Icon(Icons.more_vert,color: Colors.white70,
                            size: SizeConfig.safeBlockHorizontal * 7,
                            ),
                          )
                        ],
                      ),
                    ),
                    Expanded(
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.start,
                        children: <Widget>[
                          Icon(
                            Icons.access_time,
                            size: SizeConfig.safeBlockHorizontal * 5,
                            color: Colors.white,
                          ),
                          SizedBox(
                            width: 10,
                          ),
                          Text(
                            widget.appointmentDate + ', ' + widget.appointmentTime,
                            style: TextStyle(
                                fontSize: SizeConfig.safeBlockHorizontal * 4.7,
                                color: Colors.white70),
                          )
                        ],
                      ),
                    )
                  ],
                ),
              ),
            ),
          
        ),
        Expanded(
          flex: 2,
          child: Container(
            decoration: BoxDecoration(
                color: Colors.white,
                borderRadius: BorderRadius.only(
                    bottomLeft: Radius.circular(25),
                    bottomRight: Radius.circular(25))),
            child: Padding(
              padding: const EdgeInsets.only(left: 20.0, right: 20.0),
              child: Column(
                children: <Widget>[
                  Expanded(
                    flex: 2,
                    child: Container(
                      width: SizeConfig.safeBlockHorizontal * 90,
                      //color: Colors.pink,
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.start,
                        children: <Widget>[
                          Expanded(
                            flex: 3,
                            child: Container(
                              width: SizeConfig.safeBlockHorizontal * 15,
                              height: SizeConfig.safeBlockVertical * 8,
                              decoration: BoxDecoration(
                                  image: DecorationImage(image: AssetImage('image/igim.jpg')),
                                  color: Colors.grey[200],
                                  borderRadius: BorderRadius.circular(16)),
                            ),
                          ),
                          SizedBox(
                            width: 10,
                          ),
                          Expanded(
                            flex: 9,
                            child: Column(
                              mainAxisAlignment: MainAxisAlignment.center,
                              crossAxisAlignment: CrossAxisAlignment.start,
                              children: <Widget>[
                                Text(
                                  widget.patienSurname,
                                  maxLines: 1,
                                  overflow: TextOverflow.ellipsis,
                                  style: TextStyle(
                                      fontSize:
                                          SizeConfig.safeBlockHorizontal * 4.7,
                                      color: Colors.black87),
                                ),
                                Text(
                                  widget.patienName,
                                  maxLines: 3,
                                  overflow: TextOverflow.ellipsis,
                                  style: TextStyle(
                                      fontSize:
                                          SizeConfig.safeBlockHorizontal * 4.7,
                                      color: Colors.black87),
                                )
                              ],
                            ),
                          ),
                          Expanded(
                            flex: 2,
                            child: GestureDetector(
                              onTap: (){
                                if(isinfoPressed == true)
                                {
                                  isinfoPressed = false;
                                  widget.onRedCloseButtonTapped();
                                  setState(() {});
                                }
                                else{
                                  isinfoPressed = true;
                                  widget.onInfoTapped();
                                  setState(() {});
                                }
                              },
                              child: Container(
                                //color: Colors.red,
                                child: isinfoPressed?Transform.rotate(
                                  angle: 0.7777,
                                  child: Icon(
                                  Icons.add_circle,
                                  size: SizeConfig.safeBlockHorizontal * 9,
                                  color: Colors.redAccent,
                                ),
                                )
                                
                                
                                :Icon(
                                  Icons.info,
                                  size: SizeConfig.safeBlockHorizontal * 9,
                                  color: Colors.blue,
                                ),
                              ),
                            ),
                          ),
                        ],
                      ),
                    ),
                  ),
                  Expanded(
                    child: Padding(
                      padding: const EdgeInsets.only(bottom: 8.0),
                      child: Container(
                        width: SizeConfig.safeBlockHorizontal * 90,
                        //color: Colors.indigo,
                        child: Row(
                          children: <Widget>[
                            Expanded(
                              flex: 3,
                              child: SizedBox(
                                height: SizeConfig.safeBlockVertical * 6,
                                child: RaisedButton(
                                  disabledElevation: 0,
                                  focusElevation: 0,
                                  highlightElevation: 0,
                                  hoverElevation: 0,
                                  elevation: 0.5,
                                  shape: RoundedRectangleBorder(
                                      borderRadius: BorderRadius.circular(50)),
                                  textColor: Colors.white,
                                  color: Colors.blue,
                                  child: Text(
                                    'accept',
                                    style: TextStyle(
                                      fontSize:
                                          SizeConfig.safeBlockHorizontal * 5.5,
                                    ),
                                  ),
                                  onPressed: widget.onAccep,
                                ),
                              ),
                            ),
                            SizedBox(
                              width: 10,
                            ),
                            Expanded(
                              flex: 2,
                              child: SizedBox(
                                height: SizeConfig.safeBlockVertical * 6,
                                child: RaisedButton(
                                  disabledElevation: 0,
                                  focusElevation: 0,
                                  highlightElevation: 0,
                                  hoverElevation: 0,
                                  elevation: 0,
                                  shape: RoundedRectangleBorder(
                                      borderRadius: BorderRadius.circular(50)),
                                  textColor: Colors.black26,
                                  color: Color(0xffEBEFFB),
                                  child: Text(
                                    'Decline',
                                    style: TextStyle(
                                      color: Color(0xff878FA6) ,
                                      fontSize:
                                          SizeConfig.safeBlockHorizontal * 5,
                                    ),
                                  ),
                                  onPressed: widget.onDecline,
                                ),
                              ),
                            ),
                          ],
                        ),
                      ),
                    ),
                  )
                ],
              ),
            ),
          ),
        )
      ],
    );
  }
}



example/lib/AppointmentCard.dart

import 'package:flutter/material.dart';
import 'package:sliding_card/sliding_card.dart';
import 'AppointmentBackCard.dart';
import 'AppointmentFrontCard.dart';
import 'sizeConfig.dart';


///This is the main Appointment card 
///it regroups the AppointmentFrontCard 
///and AppointmentBacktCard together 
///using my dart package ''SlidingCard''
class AppointmentCard extends StatefulWidget {
  const AppointmentCard({
    Key key,
    this.slidingCardController , @required this.onCardTapped
  }) : super(key: key);
 
  final  SlidingCardController slidingCardController;
  final Function onCardTapped;

  @override
  _AppointmentCardState createState() => _AppointmentCardState();
}

class _AppointmentCardState extends State<AppointmentCard> {
  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: (){
        print('Well the card was tapped');
        widget.onCardTapped();
      },
          child: SlidingCard(
        slimeCardElevation: 0.5,
       // slidingAnimationReverseCurve: Curves.bounceInOut,
        cardsGap: SizeConfig.safeBlockVertical,
        controller: widget.slidingCardController,
        slidingCardWidth: SizeConfig.horizontalBloc * 90,
        visibleCardHeight: SizeConfig.safeBlockVertical * 27,
        hiddenCardHeight: SizeConfig.safeBlockVertical * 15,
        frontCardWidget: AppointmentFrontCard(
          imgLink: 'not needed anymore',
          patienName: '@emileCode',
          patienSurname: 'on Instagram',
          appointmentDate: '12 Jan 2020 ',
          appointmentTime: '12Am',
          onInfoTapped: () {
            print('info pressed');
            widget.slidingCardController.expandCard();
          },
          onDecline: () {
            print('Declined');
          },
          onAccep: () {
            print('Accepted');
          },
          onRedCloseButtonTapped: (){
            widget.slidingCardController.collapseCard();
          },
        ),
        backCardWidget:AppointmentBackCard(
          onPhoneTapped: (){print('Phone tapped');},
          patientComment: 'Have any question? you can DM me on instagram or maybe call me'
        ),
      ),
    );
  }
}



